inc/amd64.cgu.hxx

```diff
@@ -0,0 +1,274 @@
+	 */
+	enum class Register : uint8_t {
+		
+	};
+
+	uint64_t imm64(uint64_t value);
+	uint64_t imm32(uint32_t value);
+	uint64_t imm16(uint16_t value);
+	/**
+	 * @brief this is short hand for modRM(reg,AddressingMode::RegisterDirect,rm)
+	 * 
+	 * @param reg reg operrand
+	 * @param rm reg/mem opperand
+	 * @return amd64 modRM byte ready for execution
+	 */
+	byte modRM(Register reg,Register rm);
+	/**
+	 * @brief
+	 * 
+	 * @param reg reg operrand
+	 * @param rm reg/mem opperand
+	 * @return amd64 modRM byte ready for execution
+	 */
+	byte modRM(Register reg,AddressingMode mod,Register rm);
+	/**
+	 * @brief
+	 * 
+	 * @param scale must be either 1,2,4 or 8 
+	 * @param index register holding the index
+	 * @param base register holding the base address
+	 * @return amd64 SIB byte ready for execution
+	 */
+	byte SIB(uint8_t scale, Register index, Register base);
+
+	namespace prefix
+	{
+		/**
+		 * @brief constructs a REX prefix
+		 * 
+		 * @param W
+		 * @param R
+		 * @param X
+		 * @param B
+		 * @return amd64 REX byte ready for execution
+		 */
+		byte REX(bool W, bool R, bool X, bool B);
+		/**
+		 * @brief initiates a 2 byte VEX escape sequence
+		 */
+		byte VEX_2byte = 0xC5;
+		/**
+		 * @brief initiates a 3 byte VEX escape sequence
+		 */
+		byte VEX_3byte = 0xC4;
+		/**
+		 * @brief initiates a 3 byte XOP escape sequence
+		 */
+		byte XOP = 0x8F;
+		namespace legacy
+		{
+			/**
+			 * @brief AMD64 specification: 
+			 `
+			 	In 64-bit mode, the prefix allows mixing of 16-bit, 32-bit, and 64-bit data on an instruction-byinstruction basis.
+			 `
+			 ... 
+			 `
+			 	In 64-bit mode, most instructions default to a 32-bit operand size. For these instructions, a REX prefix
+			 	can specify a 64-bit operand size, and a 66h prefix specifies a 16-bit operand size. The REX
+			 	prefix takes precedence over the 66h prefix. However, if an instruction defaults to a 64-bit operand
+			 	size, it does not need a REX prefix and it can only be overridden to a 16-bit operand size. It cannot be
+			 	overridden to a 32-bit operand size, because there is no 32-bit operand-size override prefix in 64-bit
+			 	mode.
+			 `
+			 ...
+			 `
+			 	Instructions that Cannot Use the Operand-Size Prefix: The operand-size prefix should be used
+			 	only with general-purpose instructions and the x87 FLDENV, FNSTENV, FNSAVE, and FRSTOR
+			 	instructions, in which the prefix selects between 16-bit and 32-bit operand size. The prefix is ignored
+			 	by all other x87 instructions and by 64-bit media floating-point (3DNow!™) instructions.
+			 	For other instructions (mostly SIMD instructions) the 66h, F2h, and F3h prefixes are used as opcode
+			 	extensions to extend the instruction encoding space in the 0Fh, 0F_38h, and 0F_3Ah opcode maps.
+			 `
+			 */
+			byte OperandSizeOverride = 0x66;
+			/**
+			 * @brief 
+			 * 
+			 */
+			byte AddressSizeOverride = 0x67;
```

Comment:
could not find anything to really put here that would be useful in a doc comment.
relevant specification pages: 51-52
[AMD64 vol3 51-52.pdf](https://github.com/user-attachments/files/15673882/AMD64.vol3.51-52.pdf)